keel-core/src/test/kotlin/com/netflix/spinnaker/keel/validators/DeliveryConfigValidatorTests.kt

```diff
@@ -0,0 +1,164 @@
+                kind = TEST_API_V1.qualify("whatever"),
+                spec = DummyResourceSpec("test", "im a twin", "keel")
+              )
+            ),
+            constraints = emptySet()
+          ),
+          SubmittedEnvironment(
+            name = "prod",
+            resources = setOf(
+              SubmittedResource(
+                kind = TEST_API_V1.qualify("whatever"),
+                spec = DummyResourceSpec("test", "im a twin", "keel")
+              )
+            ),
+            constraints = emptySet()
+          )
+        )
+      )
+
+      test("an error is thrown and config is deleted") {
+        expectCatching {
+          subject.validate(submittedConfig)
+        }.isFailure()
+          .isA<DuplicateResourceIdException>()
+      }
+    }
+
+    context("a delivery config with non-unique artifact references errors while persisting") {
+      // Two different artifacts with the same reference
+      val artifacts = setOf(
+        DockerArtifact(name = "org/thing-1", deliveryConfigName = configName, reference = "thing"),
+        DockerArtifact(name = "org/thing-2", deliveryConfigName = configName, reference = "thing")
+      )
+
+      val submittedConfig = SubmittedDeliveryConfig(
+        name = configName,
+        application = "keel",
+        serviceAccount = "keel@spinnaker",
+        artifacts = artifacts,
+        environments = setOf(
+          SubmittedEnvironment(
+            name = "test",
+            resources = setOf(
+              SubmittedResource(
+                metadata = mapOf("serviceAccount" to "keel@spinnaker"),
+                kind = TEST_API_V1.qualify("whatever"),
+                spec = DummyResourceSpec(data = "o hai")
+              )
+            ),
+            constraints = emptySet()
+          )
+        )
+      )
+      test("an error is thrown and config is deleted") {
+        expectCatching {
+          subject.validate(submittedConfig)
+        }.isFailure()
+          .isA<DuplicateArtifactReferenceException>()
+      }
+    }
+
+    context("a second delivery config for an app fails to persist") {
+      val submittedConfig1 = SubmittedDeliveryConfig(
+        name = configName,
+        application = "keel",
+        serviceAccount = "keel@spinnaker",
+        artifacts = setOf(DockerArtifact(name = "org/thing-1", deliveryConfigName = configName, reference = "thing")),
+        environments = setOf(
+          SubmittedEnvironment(
+            name = "test",
+            resources = setOf(
+              SubmittedResource(
+                metadata = mapOf("serviceAccount" to "keel@spinnaker"),
+                kind = TEST_API_V1.qualify("whatever"),
+                spec = DummyResourceSpec(data = "o hai")
+              )
+            ),
+            constraints = emptySet()
+          )
+        )
+      )
+
+      val submittedConfig2 = submittedConfig1.copy(name = "double-trouble")
+//      test("an error is thrown and config is not persisted") {
```

Comment:
This is a scenario that cannot happen anymore -- since we are returning bad request even before reaching `upsert`, we cannot return `TooManyDeliveryConfigsException`.

Reply:
Agreed. Just remove the test.